[FEATURE]: toggle skipping project with 'pom' packaging

kumadee opened the pull request:
## Description

We currently generate bindings for multiple languages. The maven projects are structured in such a way that the packaging type cannot be `jar`, `war` or `ear` for non-java languages.
Since, the updated version of maven install plugin insists on having a packaging type as `pom` for such projects, we feel that this plugin could provide a toggle to allow executing protoc build for such projects.

## Changes Made

- I have added a boolean configuration parameter `skipPomPackaging` to toggle whether the plugin should skip the protoc build execution or not. By default, it is set to true, which preserves the existing behavior.
- I have added a test for this feature.
- I have also updated the documentation content used by maven-site-plugin. The plugin artifactId was incorrect in `index.apt`.

## Definition of Done

Before submitting this pull request, please ensure that the following criteria have been met:

- [ ] All automated tests have passed successfully.
- [x] All manual tests have passed successfully.
- [ ] Code has been reviewed by at least one other team member.
- [ ] Code has been properly documented and commented as needed.
- [ ] All new and existing code adheres to our project's coding standards.
- [ ] All dependencies have been added or removed from the project's README or other documentation as needed.
- [ ] Any relevant documentation or help files have been updated to reflect the changes made in this pull request.
- [ ] Any necessary database migrations have been run.
- [ ] Any relevant UI changes have been reviewed and approved by the UI/UX team.

## Additional Notes

[Add any additional notes or context for the reviewer or future maintainers of this code.]

Thank you for submitting!

ee08b397 commented:
@kumadee Thanks for the pull request! Could you keep this PR small and undo the whitespace and renaming changes? 

1. We can have a separate PR later, but it will make the code review easier. 
1. Good catch on the artifactId. Renaming the Artifact is a big deal. So we want to carefully roll it out.

ee08b397 commented on protoc-maven-plugin/src/main/java/io/github/blackrock/protocjar/maven/ProtocJarMojo.java:
```diff
@@ -258,6 +258,15 @@ public class ProtocJarMojo extends AbstractMojo
+    /**
+     * This parameter (when set to false) allows to execute the protoc code generation
+     * when Maven project packaging type is POM. This is required when your packaging type cannot
+     * be one of the standard types like jar, war, ear. This usually happens for non-Java
```
Can you check if the indentation is off for this new block? It looks it can be shifted right to me. I assume the tabs vs spaces caused this issue.

kumadee replied:
should be fixed now.

ee08b397 commented on protoc-maven-plugin/src/main/java/io/github/blackrock/protocjar/maven/ProtocJarMojo.java:
```diff
@@ -282,11 +291,11 @@ public class ProtocJarMojo extends AbstractMojo
 
 	public void execute() throws MojoExecutionException {
-		if (project.getPackaging() != null && "pom".equals(project.getPackaging().toLowerCase())) {
+		if (skipPomPackaging && project.getPackaging() != null && "pom".equals(project.getPackaging().toLowerCase())) {
```
@spensonshih Separate from @kumadee ’s PR, I noticed our Null check here is old-school and verbose, probably inherited from the original codebase. 

`”pom”.equalsIgnoreCase(project.getPackaging())` should replace the existing `project.getPackaging() != null &&  ”pom”.equals(project.getPackaging().toLowerCase())`. 

For example, `”pom”.equalsIgnoreCase(null) == false` and doesn’t throw null pointer exceptions. 

Reference from `maven-javadoc-plugin`
https://github.com/apache/maven-javadoc-plugin/blob/f9d3227cd18ee61d14ea6334ddcc10d97b7e1300/src/main/java/org/apache/maven/plugins/javadoc/AbstractFixJavadocMojo.java#L786
```
if ("pom".equalsIgnoreCase(project.getPackaging())) {
    getLog().warn("This project has 'pom' packaging, no Java sources is available.");
    return null;
}
```

kumadee replied:
I felt the urge to update the existing equality check, but deferred it for another PR. If the code owners insist on refactoring, I don't mind including this requested change.

spensonshih replied:
@ee08b397 , agreed.  Let's use `equalsIgnoreCase` check here.

@kumadee feel free to include this equality check refactor in this pull-request.

kumadee replied:
fixed.

ee08b397 commented on protoc-maven-plugin/src/main/java/io/github/blackrock/protocjar/maven/ProtocJarMojo.java:
```diff
@@ -282,11 +291,11 @@ public class ProtocJarMojo extends AbstractMojo
 
 	public void execute() throws MojoExecutionException {
-		if (project.getPackaging() != null && "pom".equals(project.getPackaging().toLowerCase())) {
+		if (skipPomPackaging && project.getPackaging() != null && "pom".equals(project.getPackaging().toLowerCase())) {
```
@spensonshih To be consistent with other popular Maven plugins, how about we implement a `@Parameter skip` and default to null, checking if it’s explicitly set to false by caller, then we don’t skip no matter packaging is POM or not. If null, keep the current behavior. If set to true, skip despite packaging type. 

The gist is I find @kumadee ‘s use case is legit but I don’t recall other Maven being plugins used in Python requirements.txt have switches named as such `skipPomPackaging`. It’s good to be consistent among the popular ones, be easy to use for future users.

kumadee replied:
Should we not preserve the existing behavior of the plugin as default? The new switch is an opt-in, whereas you suggest it to be opt-out.

spensonshih replied:
@kumadee I did find the use of `pom` packaging a bit strange, when your project is actually generating .py files.  In this case, shouldn't you use a custom packaging type rather than `pom`?  See https://www.baeldung.com/maven-packaging-types

kumadee replied:
@spensonshih Custom packaging type cannot be used in our case. The referred blogpost explains how to support custom packaging types generated by a maven plugin.
In our case, we compile the proto to py source and then use maven-assembly-plugin to package it to zip. The `maven-assembly-plugin` only allows packaging to multiple formats but doesn't extend the packaging attribute of a pom.

spensonshih replied:
@kumadee , I’m still not convinced that you should use pom packaging type.  Since your project is generating py files and packaged as a zip file via the maven-assembly-plugin.  Using pom packaging type indicates your project is an aggregator or a parent.  I’m worried a future Maven version could skip some of the lifecycle phases for pom packaging types.

Can you share why the project cannot use jar packaging type?  I was able to create a simple test using jar packaging and running maven-assembly-plugin to generate the zip at the same time.

ee08b397 replied:
@kumadee Updated my first comment on `skip`, sorry about confusion. It would be great if you could link your GitHub repo and usage here in documentation, which is an awesome user story to tell!



@spensonshih Good idea on showing Python usage. We want to understand user requirements deeply. 
My understanding of @kumadee’s need is `maven-install-plugin` v3.0.0+ enforcing `packaging:pom` if there's no Java artifact generated. Given Python source code is generated here, it would throw an error like “The packaging for this project did not assign a file to the build artifact” if `<packaging>` is not `pom`, such as `JAR` or any custom type. See the sample `/non-java-test/pom.xml`.

ee08b397 reviewed (changes requested):
Thanks for the contribution and extending it to work with Python! Made some suggestions.

kumadee committed:
feat: toggle skipping project with 'pom' packaging

Additionally:
- includes test for the new configuration parameter.
- correction of plugin artifactId in documentation. It is now inline with the released names in maven central.

ee08b397 commented on protoc-maven-plugin/src/main/java/io/github/blackrock/protocjar/maven/ProtocJarMojo.java:
```diff
@@ -258,6 +258,15 @@ public class ProtocJarMojo extends AbstractMojo
 	private String protocArtifact;
 
+	/**
+	* This parameter (when set to false) allows to execute the protoc code generation
```
I also suggest we simplify the JavaDoc and make it easier to follow. Such as “Set parameter to false from default true to override the default ‘pom’ packaging skipping logic, when working with non-Java languages. The default logic is required for standard Java packaging like JAR, WAR and EAR“.

kumadee commented:
> @kumadee Thanks for the pull request! Could you keep this PR small and undo the whitespace and renaming changes?
> 
> 1. We can have a separate PR later, but it will make the code review easier.

My bad. I did not notice tabs vs whitespaces here. It's reverted now.

> 2. Good catch on the artifactId. Renaming the Artifact is a big deal. So we want to carefully roll it out.

I am not renaming the artifactId now. It was already renamed with the first release of this fork. I made the site documentation inline with the README (https://github.com/blackrock/protoc-jar-maven-plugin?tab=readme-ov-file#usage) and what is published on maven central.

kumadee committed:
refactor: project packaging type check

ee08b397 commented:
@kumadee Friendly reminder: Could you provide context for your Python project using this Java Maven POM packaging?  We are looking forward to a use case like this. Thanks!

kumadee commented:
Sorry folks. I am still on a long vacation and forgot to address this PR before leaving.

I will reply to the question by next week. Thanks for your patience.

kumadee commented:
I have closed the PR because we had to move another plugin to compile protobufs for edition 4.30.x or higher.
Thanks for your patience.
